ipywidgets/widgets/trait_types/eventful.py

```diff
@@ -0,0 +1,180 @@
+
+
+class Callback(object):
+
+    def __init__(self, owner, trait, etype, callback):
+        self.owner = owner
+        self.trait = trait
+        self.etype = etype
+        self.func = callback
+
+    def __eq__(self, other):
+        if not isinstance(other, Callback):
+            return self.func == other
+        else:
+            return other is self
+
+
+class Beforeback(Callback):
+
+    def __call__(self, inst, call):
+        call = call.copy()
+        call.update(
+            trait=self.trait,
+            owner=self.owner,
+            type=self.etype)
+        return self.func(inst, call)
+
+
+class Afterback(Callback):
+
+    def __call__(self, inst, answer):
+        if self.func is None:
+            result = answer['before']
+            if (isinstance(result, Redirect)
+                and result.origin is not None):
+                origin = result.origin
+            else:
+                origin = self.etype
+
+            event = (result(answer.value) if
+                callable(result) else result)
+        else:
+            origin = self.etype
+            event = self.func(inst, answer)
+        if None not in (origin, event):
+            self.owner.notify_change(Bunch(
+                type=origin, event=event,
+                name=self.trait.name,
+                owner=self.owner))
+        return event
+
+
+class Redirect(object):
+
+    def __init__(self, origin, target, inst, args, kwargs):
+        self.origin = origin
+        registry = inst._instance_spectator._callback_registry
+        # Eventful._setup_events registers first
+        b, a = registry.get(target, [(None, None)])[0]
+
+        if b is not None:
+            call = Bunch(name=target,
+                kwargs=kwargs.copy(),
+                args=args[:])
+            bval = b(inst, call)
+        else:
+            bval = None
+
+        if a is not None:
+            self.trigger = lambda value: a(inst, Bunch(
+                before=bval, name=target, value=value))
+        elif callable(bval):
+            self.trigger = bval
+
+    def __call__(self, value):
+        # trigger afterbacks
+        return self.trigger(value)
+
+    @staticmethod
+    def trigger(value):
+        pass
+
+
+class Eventful(TraitType):
+
+    klass = None
+    event_map = {}
+    type_name = None
+    watchable_type = None
+
+    def __init__(self, *args, **kwargs):
+        super(Eventful, self).__init__(*args, **kwargs)
+        default_value = self.default()
+        if self.klass is None and default_value is not Undefined:
+            self.klass = type(default_value)
+        self._active_events = []
+        self.setup_events()
+
+    def event(self, type, on, before=None, after=None):
+        for method in (on if isinstance(on, (list, tuple)) else (on,)):
+            self._active_events.append((type, method, before, after))
+        return self
+
+    def setup_events(self):
+        for name, on in self.event_map.items():
```

Comment:
Maps arbitrary names (used for the `_before_<name>` and `_after_<name>` convention) to methods.